File: randology-dump-master/thesis/src/uk/ac/stir/randology/generator/Webkit2.java
package uk.ac.stir.randology.generator;

// This quadratic generator is from Webkit2, see https://gist.github.com/Protonk/5367430
public class Webkit2 implements RNG {

	private long state;
	
	
	public Webkit2(long seed) {
		state = seed;
	}
	
	public long nextLong() {
		state += (state * state) ^ 5;
		return Math.abs(state);
	}
	
    public long next64bits() {
    	return nextLong();
    }
	
	public double nextDouble() {
	    return Math.abs( (double)nextLong() / Long.MAX_VALUE );
	}
	
	@Override
	public String getName() {
		return "Webkit2";
	}

}

File: python-implementation/src/pnrg/CSG/Webkit2.py
import numpy as np

from pnrg import RNG


# This quadratic generator is from Webkit2, see https://gist.github.com/Protonk/5367430
class Webkit2(RNG):
    def __init__(self, seed):
        super().__init__("Webkit2")
        self.state = np.int64(seed)

    def next_int(self) -> np.int64:
        self.state += (self.state * self.state) ** 5
        return abs(self.state)

    def next_float(self) -> np.float64:
        return abs(self.next_int() / np.iinfo(np.int64).max)

    def next_64_bits(self) -> np.int64:
        return self.next_int()
